### PR TITLE
parallel version of mergesort is created

### PR DESCRIPTION
**Pull request by anishsingh935**
#### Checklist

- [ ] Added description of change
- [ ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [ ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: graphics/spirograph_0.3_0.75_10.csv [hunk 1]
```diff
@@ -0,0 +1,500 @@
+0.475,0.475
```
Comment (ayaankhan98) [inline graphics/spirograph_0.3_0.75_10.csv:1]: please remove this file.

File: sorting/merge_sort.cpp [hunk 2]
```diff
@@ -14,6 +14,7 @@
  *
  */
 #include <iostream>
+using namespace std;
```
Comment (ayaankhan98) [inline sorting/merge_sort.cpp:17]: please undo the changes of this file.

File: graphics/spirograph_0.3_0.75_10.csv (hunk 1 above)
Comment (anishsingh935) [reply graphics/spirograph_0.3_0.75_10.csv:1]: Is there any problem

Comment (anishsingh935) [reply graphics/spirograph_0.3_0.75_10.csv:1]: please suggest me that where I have made mistake

**Comment by anishsingh935** [comment]
//merge sort parallel
//compile with -fopenmp flag 
//to calculate time taken on linux, use "time" keyword while executing object file

#include<iostream>
#include "omp.h"

//performs the bottom up combine operation
void merge(int a[], int l, int m, int r)
{
	int temp[m-l+1], temp2[r-m];
	//copy first part to temporary array
	for(int i=0; i<(m-l+1); i++){
		temp[i]=a[l+i];}
	//copy second part to temporary arrat]y
	for(int i=0; i<(r-m); i++)
	{	temp2[i]=a[m+1+i];}
	int i=0, j=0, k=l;
	//combine the arrays on the basis of order
	while(i<(m-l+1) && j<(r-m))
	{
		if(temp[i]<temp2[j]){
			a[k++]=temp[i++];}
		else{
			a[k++]=temp2[j++];
	}

	}
	//to combine the remainder of the two arrays
	while(i<(m-l+1))
	{	a[k++]=temp[i++];}
	while(j<(r-m))
		{a[k++]=temp2[j++];}

}
/* 
This function is the main hotspot of the algorithm which I have chosen to parallelise
The calls labelled call 1 and call 2 are essentially independent since they are operating on distinct
parts of the array and hence can be called in parallel.
*/

void mergeSort(int a[], int l, int r)
{
	if(l<r)
	{
		//cout<<"came here";
		int m=(l+r)/2;
		#pragma omp parallel sections num_threads(2)
		{
			#pragma omp section
			{
				mergeSort(a,l,m); //call 1
			}
			#pragma omp section
			{
				mergeSort(a,m+1,r); //call 2
			}
		}
		merge(a,l,m,r); //this function is outside the parallel sections since it merges and access the same data 
	}
}
//a simple print function to print all the elements
void print(int a[], int n)
{
	for(int i=0; i<n; i++){
		std::cout<<a[i]<<" ";}
	std::cout<<std::endl;
}
int main()
{
	int n;
    std::cout<<" Enter the size of the array"<<std::endl;
	std::cin>>n;
	int a[n];
    std::cout<<"Enter the element of tha array"<<std::endl;
	for(int i=0; i<n; i++)
		{std::cin>>a[i];}
	std::cout<<"\nUnsorted array; ";
	print(a,n);
	//call should be to a wrapper function that calls the mergeSort function
	mergeSort(a,0,n-1);
	std::cout<<"\nSorted array: ";
	print(a,n);
	return 0;
}

**Comment by anishsingh935** [comment]
This is my code of parallel version of merge sort

File: graphics/spirograph_0.3_0.75_10.csv (hunk 1 above)
Comment (ayaankhan98) [reply graphics/spirograph_0.3_0.75_10.csv:1]: that file is not related to this PR.

**Review by ayaankhan98 — COMMENTED** [review]
please read the [repository standard guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md) and follow

File: sorting/merge_sort.cpp (hunk 2 above)
Comment (anishsingh935) [reply sorting/merge_sort.cpp:17]: ok

**Comment by faizan2700** [comment]
This repo basically divide the given array to be sorted by merge sort in two parts and schedule two different threads to sort each half and after that sequential merge is used. I don't think it is best utilization of parallel computing because threads keep changing. A better approach would be to run it sequentially till the array is divided into parts equal to processors available (array is divided into 8 parts and 8 processors are available for example. ) And then schedule each part on different processor and then sequentially merge them. @ayaankhan98 , @Panquesito7 , @Nirzak what do you think about this?
